Use gsutil to upload test artifacts and console log to GCS

ixdy opened the pull request:
Automatically upload the console log and test artifacts and set a "latest build" file in GCS for all gce and gke-ci builds. At least partway fixes #13548.

This currently only supports the post-submit Jenkins projects; I need to figure out the best way to trigger this if something fails in build.sh as well.

ixdy committed:
Add script to upload test artifacts and log to GCS using gsutil

This new script is designed to be source'd by Jenkins jobs. It'll
automatically upload all test artifacts and the console log to GCS on
exit.

ixdy commented:
Should be ready for review now. @kubernetes/goog-testing

ixdy commented:
https://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce/9925/ is a run where it actually uploaded things.

lavalamp commented on hack/jenkins/e2e.sh:
```diff
@@ -430,6 +469,8 @@ case ${JOB_NAME} in
           ${GCE_DEFAULT_SKIP_TESTS[@]:+${GCE_DEFAULT_SKIP_TESTS[@]}} \
           ${GCE_FLAKY_TESTS[@]:+${GCE_FLAKY_TESTS[@]}} \
           )"}
+    : ${JENKINS_GCS_LOGS_PATH:="gs://kubernetes-jenkins/logs"}
```
Why don't we want this on for all of the tests?

ixdy replied:
wasn't sure if we also wanted all GKE builds uploaded, but I guess it's harmless enough.

lavalamp commented on hack/jenkins/e2e.sh:
```diff
@@ -28,6 +28,36 @@ set -o nounset
+      gsutil -m -q -o "GSUtil:use_magicfile=True" cp -a "${gcs_acl}" -r -c \
+        -z log,txt,xml "${ARTIFACTS}" "${gcs_build_path}/artifacts" || true
+    fi
+    # Remove ANSI escape sequences from the console log before uploading.
```
I thought there was a setting to not output these in the first place-- can we turn that on instead?

ixdy replied:
we have a plugin which renders the colors on Jenkins, which is kinda nice.

lavalamp commented on hack/jenkins/e2e.sh:
```diff
@@ -28,6 +28,36 @@ set -o nounset
+    echo "Uploading to ${gcs_build_path}"
+    if [[ -n "${ARTIFACTS:-}" ]]; then
+      gsutil -m -q -o "GSUtil:use_magicfile=True" cp -a "${gcs_acl}" -r -c \
+        -z log,txt,xml "${ARTIFACTS}" "${gcs_build_path}/artifacts" || true
```
concerned about the || true, seems like we will have a hard time noticing if we skip things?

quinton-hoole replied:
See also https://github.com/kubernetes/kubernetes/issues/7778

ixdy replied:
This was intentional - we don't want failing gsutil uploads to fail the entire test run; see #13548.

ixdy replied:
added a comment noting such

quinton-hoole commented:
@ixdy Needs a rebase.

ixdy commented:
rebased and comments addressed, PTAL

ixdy commented:
Actually, I may have a way of making this work for all of our builds, not just those that call e2e.sh. Standby for updates...

ixdy commented:
OK, I've updated this PR with a new script rather than changing e2e.sh.

The basic usage is that we add another (series of) command(s) to all of the Jenkins jobs:
```shell
mkdir -p _tmp/ && curl -fsS --retry 3 -o _tmp/upload-to-gcs.sh "https://raw.githubusercontent.com/kubernetes/kubernetes/master/hack/jenkins/upload-to-gcs.sh" && source _tmp/upload-to-gcs.sh
```

Basically, we download this script and source it into the bash session running the test commands. This creates the function and installs a trap to automatically upload logs to GCS on exit.

We can thus use this with all of the Jenkins jobs, not just those running e2e.sh.

I also fixed the git commit environment variable.

ixdy commented:
@k8s-bot test this please

ixdy commented:
Improvements this has over the GCS Jenkins plugin:
* all files are gzipped before upload, which means they upload faster and use less space on GCS. (They're transparently unzipped when downloaded.)
* artifacts are uploaded in parallel rather than serially (should be faster esp. for builds like the scalability project)
* all files get a proper content-type set, rather than using the generic default
* all files can be viewed directly in the browser, rather than having to be downloaded first (e.g. compare [build-log.txt](https://pantheon.corp.google.com/m/cloudstorage/b/kubernetes-jenkins/o/pr-logs/7c3434eca324bd212214cfae865209f488981d41/kubernetes-pull-build-test-e2e-gce/11816/build-log.txt), uploaded by the Jenkins plugin, vs [console-log.txt](https://pantheon.corp.google.com/m/cloudstorage/b/kubernetes-jenkins/o/pr-logs/7c3434eca324bd212214cfae865209f488981d41/kubernetes-pull-build-test-e2e-gce/11816/console-log.txt), uploaded by this script)
* the latest-build.txt file can be watched by other jobs to determine when a new build has completed

wojtek-t commented:
Nice - waiting for this to be in.

wojtek-t commented on hack/jenkins/upload-to-gcs.sh:
```diff
@@ -0,0 +1,75 @@
+    local -r gcs_acl="public-read"
+    echo "Uploading to ${gcs_build_path}"
+    if [[ -d "${artifacts_path}" ]]; then
+      gsutil -m -q -o "GSUtil:use_magicfile=True" cp -a "${gcs_acl}" -r -c \
```
Can we retry uploading in case of failure (e.g. 3 times)? This will make it more robust.

Same in other places.

ixdy replied:
added retry loop. waiting for kubernetes-test-go to run again to see if it worked.

wojtek-t commented on hack/jenkins/upload-to-gcs.sh:
```diff
@@ -0,0 +1,75 @@
+# JENKINS_UPLOAD_TO_GCS is set to y.
+# We intentionally ignore gsutil errors since we don't want failed uploads to
+# fail the entire test run (#13548).
+function upload_logs_to_gcs() {
```
is this references from somewhere?

krousey replied:
Referenced in the trap command at the bottom. Jenkins will source this in and that trap command will execute it. See https://github.com/kubernetes/kubernetes/pull/14454#issuecomment-146390939 for more details.

krousey commented:
@ixdy This looks good to me (as far as bash looks good). Care to comment on https://github.com/kubernetes/kubernetes/pull/14454#discussion_r41750878 ?

krousey commented:
@ixdy This gets my LGTM. Feel free to add it when you're satisfied with the test results.

wojtek-t commented on hack/jenkins/upload-to-gcs.sh:
```diff
@@ -0,0 +1,79 @@
+  done
+}
+
+
```
nit: remove one empty line

ixdy replied:
done

wojtek-t commented:
lgtm

ixdy committed:
Remove extraneous newline

ixdy commented:
After updating the Jenkins job to actually download from the correct branch (oops), it seems to be working correctly.

ixdy commented:
self-added lgtm label per @krousey in https://github.com/kubernetes/kubernetes/pull/14454#issuecomment-148226816
